### ISSUE TITLE
diagnostic for redeclarations

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ty#2987]
```py
def foo(): ...
def foo(): ... # no error

bar: int = 1
bar: int = 2 # no error
```

basically a better version of pyright's `reportRedeclaration` rule ([more info on its issues here](https://docs.basedpyright.com/latest/benefits-over-pyright/fixes-for-rules/#reportredeclaration-and-reportduplicateimport))

**Comment by carljm** [comment]
Re-declaration (shadowing) is explicitly and intentionally supported in ty; that's a feature.

That said, for those who want to enforce a style that prohibits it, we could add an opt-in lint rule for that.

**Comment by DetachHead** [comment]
shadowing isn't necessarily the same as redeclaration as i understand it. [according to wikipedia](https://en.wikipedia.org/wiki/Variable_shadowing), shadowing is a redeclaration *in a different scope*, and i think that distinction is important to make, since shadowing is more likely to be intentional than redeclaration IMO.

perhaps there could be two separate rules `redeclaration` and `shadowed-name` so that users can decide for themselves which kind of redeclarations they want to allow. i personally like to ban both in my projects, but that's just my personal preference

**Comment by DetachHead** [comment]
i just realized ruff's [`redefined-while-unused` (`F811`)](https://docs.astral.sh/ruff/rules/redefined-while-unused/) partially covers redeclarations, so perhaps this could be a ruff rule instead.

i also found https://github.com/astral-sh/ruff/issues/10522 which should fully cover function redeclarations. i also opened https://github.com/astral-sh/ruff/issues/23802 since it doesn't seem to work on variables

**Comment by carljm** [comment]
The term "shadowing" is commonly used in Rust at least to refer to a new `let` for the same variable name, in the same scope. I guess that usage disagrees with Wikipedia.

But either way, I agree those should be separate lint rules. There's also https://github.com/astral-sh/ruff/pull/23305 pending in ty.
